### PR TITLE
Add Laplace distribution

### PR DESCRIPTION
**Pull request by YoshikawaMasashi**
split from #4678 .

### CODE REVIEW

File: chainer/distributions/laplace.py [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+
+
+class Laplace(distribution.Distribution):
+
+    """Laplace Distribution.
+
+    The probability density function of the distribution is expressed as
+
+    .. math::
+        p(x;\\mu,\\sigma) = \\frac{1}{2b}
```
Comment (toslunar) [inline chainer/distributions/laplace.py]: Fix typo: `\\sigma` -> `b`.

File: chainer/distributions/laplace.py [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+        scale :math:`b`.
+    """
+
+    def __init__(self, loc, scale):
+        super(Laplace, self).__init__()
+        self.loc = chainer.as_variable(loc)
+        self.scale = chainer.as_variable(scale)
+
+    def __copy__(self):
+        return self._copy_to(Laplace(self.loc, self.scale))
```
Comment (toslunar) [inline chainer/distributions/laplace.py]: Could you explain `__copy__` or `_copy_to` in the docstring of `chainer.Distribution`?

File: chainer/distributions/laplace.py [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+        return self.loc
+
+    @property
+    def mode(self):
+        return self.loc
+
+    def prob(self, x):
+        bl = broadcast.broadcast_to(self.loc, x.shape)
+        bs = broadcast.broadcast_to(self.scale, x.shape)
+        return 0.5 / bs * exponential.exp(- basic_math.absolute(x - bl) / bs)
```
Comment (toslunar) [inline chainer/distributions/laplace.py]: Why don't you use `abs(x - bl)`?

File: chainer/distributions/laplace.py [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+    def batch_shape(self):
+        return self.loc.shape
+
+    def cdf(self, x):
+        bl = broadcast.broadcast_to(self.loc, x.shape)
+        bs = broadcast.broadcast_to(self.scale, x.shape)
+        return clip.clip(0.5 * exponential.exp(
+            (x - bl) / bs), 0., 0.5) \
+            + clip.clip(0.5 - 0.5 * exponential.exp(
+                -(x - bl) / bs), 0., 0.5)
```
Comment (toslunar) [inline chainer/distributions/laplace.py]: This formula is non-differentiable at `x == self.loc`.  How about implementing
`class LaplaceCDF(FunctionNode)` whose backward returns `gy * pdf` for `gx`.

File: chainer/distributions/laplace.py [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+                size=(n,) + self.loc.shape).astype(numpy.float32)
+
+        noise = broadcast.broadcast_to(self.scale, eps.shape) * eps
+        noise += broadcast.broadcast_to(self.loc, eps.shape)
+
+        return noise
+
+    @property
+    def stddev(self):
+        return sqrt.sqrt(2 * self.scale ** 2)
```
Comment (toslunar) [inline chainer/distributions/laplace.py]: Can we assume `self.scale` is positive?  Then use `numpy.sqrt(2) * self.scale`.

File: chainer/distributions/laplace.py (hunk 4 above)
Comment (toslunar) [reply chainer/distributions/laplace.py]: On this point could you give me time to try implementing it?

Comment (YoshikawaMasashi) [reply chainer/distributions/laplace.py]: OK. Thanks!

File: chainer/distributions/laplace.py (hunk 3 above)
Comment (YoshikawaMasashi) [reply chainer/distributions/laplace.py]: I mistaked.

File: chainer/distributions/laplace.py (hunk 2 above)
Comment (YoshikawaMasashi) [reply chainer/distributions/laplace.py]: Sorry. it is not needed.

File: tests/chainer_tests/distributions_tests/test_laplace.py [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+
+
+@testing.parameterize(*testing.product({
+    'shape': [(2, 3), ()],
+    'dtype': [numpy.float32, numpy.float64],
+}))
+class TestLaplaceICDF(unittest.TestCase):
+
+    def setUp(self):
+        self.x = numpy.random.uniform(size=self.shape).astype(self.dtype)
```
Comment (toslunar) [inline tests/chainer_tests/distributions_tests/test_laplace.py]: I'm sorry.  The backward test seems unstable near 0 or 1.  Could you fix it?  How about passing `low=0.02, high=0.98` to `numpy.random.uniform`?

Comment (YoshikawaMasashi) [reply tests/chainer_tests/distributions_tests/test_laplace.py]: OK

**Comment by toslunar** [comment]
Jenkins, test this please.

**Comment by toslunar** [comment]
The Jenkins failure is irrelevant to this PR.

**Pull request merged**